File: samples/tutorials-native/tutorials-3d/tutorial-3/java/src/main/java/com/scichart/tutorial/MainActivity.java
package com.scichart.tutorial;

import android.os.Bundle;
import android.util.Log;
import android.widget.LinearLayout;

import androidx.appcompat.app.AppCompatActivity;

import com.scichart.charting.visuals.SciChartSurface;
import com.scichart.charting3d.model.dataSeries.xyz.XyzDataSeries3D;
import com.scichart.charting3d.modifiers.OrbitModifier3D;
import com.scichart.charting3d.modifiers.PinchZoomModifier3D;
import com.scichart.charting3d.modifiers.TooltipModifier3D;
import com.scichart.charting3d.modifiers.ZoomExtentsModifier3D;
import com.scichart.charting3d.visuals.SciChartSurface3D;
import com.scichart.charting3d.visuals.axes.NumericAxis3D;
import com.scichart.charting3d.visuals.pointMarkers.SpherePointMarker3D;
import com.scichart.charting3d.visuals.renderableSeries.scatter.ScatterRenderableSeries3D;
import com.scichart.core.framework.UpdateSuspender;

import java.util.Collections;
import java.util.Random;

public class MainActivity extends AppCompatActivity {

    private SciChartSurface3D surface;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        try {
            SciChartSurface.setRuntimeLicenseKey("");
        } catch (Exception e) {
            Log.e("SciChart", "Error during setting license key", e);
        }

        final LinearLayout chartLayout = findViewById(R.id.chart_layout);

        surface = new SciChartSurface3D(this);
        chartLayout.addView(surface);

        final XyzDataSeries3D dataSeries = new XyzDataSeries3D(Double.class, Double.class, Double.class);
        for (int i = 0; i < 200; i++) {
            final double x = getGaussianRandomNumber(5.0, 1.5);
            final double y = getGaussianRandomNumber(5.0, 1.5);
            final double z = getGaussianRandomNumber(5.0, 1.5);
            dataSeries.append(x, y, z);
        }

        final SpherePointMarker3D pointMarker = new SpherePointMarker3D();
        pointMarker.setFill(0xFF32CD32);
        pointMarker.setSize(10.0f);

        final ScatterRenderableSeries3D rSeries = new ScatterRenderableSeries3D();
        rSeries.setDataSeries(dataSeries);
        rSeries.setPointMarker(pointMarker);

        // <CreateTooltipModifier>
        final TooltipModifier3D tooltipModifier = new TooltipModifier3D();
        tooltipModifier.setExecuteOnPointerCount(2);
        // </CreateTooltipModifier>

        UpdateSuspender.using(surface, () -> {
            surface.setXAxis(new NumericAxis3D());
            surface.setYAxis(new NumericAxis3D());
            surface.setZAxis(new NumericAxis3D());
            Collections.addAll(surface.getRenderableSeries(), rSeries);

            // <AddModifiers>
            Collections.addAll(surface.getChartModifiers(),
                    new OrbitModifier3D(),
                    new ZoomExtentsModifier3D(),
                    new PinchZoomModifier3D(),
                    tooltipModifier
            );
            // </AddModifiers>
        });
    }

    private Double getGaussianRandomNumber(Double mean, Double stdDev) {
        final Random random = new Random();

        final double u1 = random.nextDouble();
        final double u2 = random.nextDouble();

        final double randStdNormal = Math.sqrt(-2.0 * Math.log(u1)) * Math.sin(2.0 * Math.PI * u2);
        return mean * stdDev * randStdNormal;
    }
}

File: samples/tutorials-native/tutorials-3d/tutorial-1/kotlin/src/main/java/com/scichart/tutorial/MainActivity.kt
package com.scichart.tutorial

import android.os.Bundle
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import com.scichart.charting.visuals.SciChartSurface
import com.scichart.charting3d.model.dataSeries.xyz.XyzDataSeries3D
import com.scichart.charting3d.visuals.SciChartSurface3D
import com.scichart.charting3d.visuals.axes.NumericAxis3D
import com.scichart.charting3d.visuals.pointMarkers.SpherePointMarker3D
import com.scichart.charting3d.visuals.renderableSeries.scatter.ScatterRenderableSeries3D
import com.scichart.core.framework.UpdateSuspender
import java.util.*
import kotlin.math.ln
import kotlin.math.sin
import kotlin.math.sqrt

class MainActivity : AppCompatActivity() {

    private lateinit var surface: SciChartSurface3D

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        SciChartSurface.setRuntimeLicenseKey("")

        val chartLayout = findViewById<ViewGroup>(R.id.chart_layout)

        surface = SciChartSurface3D(this)
        chartLayout.addView(surface)

        // <CreateDataSeries>
        val dataSeries = XyzDataSeries3D(Double::class.javaObjectType, Double::class.javaObjectType, Double::class.javaObjectType)
        for (i in 0..200) {
            val x = getGaussianRandomNumber(5.0, 1.5)
            val y = getGaussianRandomNumber(5.0, 1.5)
            val z = getGaussianRandomNumber(5.0, 1.5)
            dataSeries.append(x, y, z)
        }
        // </CreateDataSeries>

        // <CreateRenderableSeries>
        val pointMarker = SpherePointMarker3D()
        pointMarker.fill = 0xFF32CD32.toInt()
        pointMarker.size = 10.0f

        val rSeries = ScatterRenderableSeries3D()
        rSeries.dataSeries = dataSeries
        rSeries.pointMarker = pointMarker
        // </CreateRenderableSeries>

        UpdateSuspender.using(surface) {
            // <AddAxes>
            surface.xAxis = NumericAxis3D()
            surface.yAxis = NumericAxis3D()
            surface.zAxis = NumericAxis3D()
            // </AddAxes>

            // <AddRenderableSeries>
            Collections.addAll(surface.renderableSeries, rSeries)
            // </AddRenderableSeries>
        }
    }

    private fun getGaussianRandomNumber(mean: Double, stdDev: Double) : Double {
        val random = Random()

        val u1: Double = random.nextDouble()
        val u2: Double = random.nextDouble()

        val randStdNormal = sqrt(-2.0 * ln(u1)) * sin(2.0 * Math.PI * u2)
        return mean * stdDev * randStdNormal
    }
}
